src/main/java/org/opentripplanner/routing/core/RoutingContext.java

```diff
@@ -399,10 +399,32 @@ public boolean isWheelchairAccessible(Vertex v) {
     }
 
     /**
-     * Tear down this routing context, removing any temporary edges.
+     * Tear down this routing context, removing any temporary edges from
+     * the "permanent" graph objects. This enables all temporary objects
+     * for garbage collection.
      */
     public void destroy() {
-        if (origin instanceof TemporaryVertex) ((TemporaryVertex) origin).dispose();
-        if (target instanceof TemporaryVertex) ((TemporaryVertex) target).dispose();
+        disposeTemporaryStart(fromVertex, null);
+        disposeTemporaryEnd(toVertex, null);
+    }
+
+    private static void disposeTemporaryStart(Vertex v, Edge incoming) {
+        if (v instanceof TemporaryVertex) {
+            for (Edge edge : v.getOutgoing()) {
+                disposeTemporaryStart(edge.getToVertex(), edge);
+            }
+        } else if (incoming != null) {
+            v.removeIncoming(incoming);
+        }
+    }
+
+    private static void disposeTemporaryEnd(Vertex v, Edge outgoing) {
```

Comment:
Once I realized how this works the parameter names became clearer. But at first I found the names outgoing/incoming strange because in the recursive call, we iterate over all outgoing edges but set a parameter called "incoming". Of course it's incoming from the point of view of the target vertex being passed in the first parameter. But this will all be much clearer to a first-time reader if it has Javadoc including `@param` annotations. I see now it's actually quite simple: we're repeatedly calling with a reachable vertex and the edge used to reach that vertex.

Reply:
@abyrd FYI: I have done this now without recursion and with loop protection. It was not that difficult. But it requires a better unit test. I have started on the test, but need a little more time on it.